File: S84.CTCode.Transpiler.StringHelper.ctcode.cpp
#include "S84.CTCode.Transpiler.StringHelper.ctcode.hpp"

namespace s84 {
namespace ctcode {
namespace transpiler {
namespace stringhelper {
namespace ctcode {

StringHelper::StringHelper()
{
}

void StringHelper::WriteLines(OmniPointer<s84::ctcode::system::ctcode::OutputStream> destination, std::vector<std::string> lines)
{
    int lines_index = 0;
    while (lines_index<Size(lines))
    {
        std::string line = Element(lines,lines_index);
        destination->WriteLine(line);
        lines_index = lines_index+1;
    }
}

std::string StringHelper::Indentation(int indent)
{
    std::string result = "";
    result = std::string("");
    while (indent>0)
    {
        indent = indent-1;
        result = Concat(result,std::string("    "));
    }
    return result;
}

bool StringHelper::BeginsWith(std::string prefix, std::string value)
{
    if (Length(prefix)>Length(value))
    {
        return false;
    }
    int prefix_index = 0;
    while (prefix_index<Length(prefix))
    {
        if (At(prefix,prefix_index)!=At(value,prefix_index))
        {
            return false;
        }
        prefix_index = prefix_index+1;
    }
    return true;
}

std::string StringHelper::StripDot(std::string input)
{
    int index = 0;
    index = 0;
    std::string result = "";
    result = std::string("");
    while (index<Length(input))
    {
        std::string character = "";
        character = At(input,index);
        if (character==std::string("."))
        {
            result = Concat(result,std::string("_"));
        }
        else
        {
            result = Concat(result,character);
        }
        index = index+1;
    }
    return result;
}

std::string StringHelper::SnakeCaseToCamelCase(std::string snake_case)
{
    bool capitalize_this_letter = true;
    std::string camel_case = std::string("");
    int index = 0;
    index = 0;
    while (index<Length(snake_case))
    {
        std::string source_character = At(snake_case,index);
        if (source_character==std::string("_"))
        {
            capitalize_this_letter = true;
        }
        else
        {
            if (capitalize_this_letter==true)
            {
                std::string upper_character = this->CharacterToUpper(source_character);
                capitalize_this_letter = false;
                camel_case = Concat(camel_case,upper_character);
            }
            else
            {
                capitalize_this_letter = false;
                camel_case = Concat(camel_case,source_character);
            }
        }
        index = index+1;
    }
    return camel_case;
}

std::string StringHelper::CamelCaseToSnakeCase(std::string camel_case)
{
    std::string delimiter = std::string("_");
    std::string snake_case = std::string("");
    int index = 0;
    while (index<Length(camel_case))
    {
        std::string source_character = At(camel_case,index);
        std::string lower_character = this->CharacterToLower(source_character);
        if (this->IsUpper(source_character)||this->IsDigit(source_character))
        {
            bool is_first_character = Length(snake_case)==0;
            if (!is_first_character)
            {
                snake_case = Concat(snake_case,delimiter);
            }
        }
        snake_case = Concat(snake_case,lower_character);
        index = index+1;
    }
    return snake_case;
}

bool StringHelper::IsUpper(std::string character)
{
    bool result = false;
    result = false;
    result = result||character==std::string("A");
    result = result||character==std::string("B");
    result = result||character==std::string("C");
    result = result||character==std::string("D");
    result = result||character==std::string("E");
    result = result||character==std::string("F");
    result = result||character==std::string("G");
    result = result||character==std::string("H");
    result = result||character==std::string("I");
    result = result||character==std::string("J");
    result = result||character==std::string("K");
    result = result||character==std::string("L");
    result = result||character==std::string("M");
    result = result||character==std::string("N");
    result = result||character==std::string("O");
    result = result||character==std::string("P");
    result = result||character==std::string("Q");
    result = result||character==std::string("R");
    result = result||character==std::string("S");
    result = result||character==std::string("T");
    result = result||character==std::string("U");
    result = result||character==std::string("V");
    result = result||character==std::string("W");
    result = result||character==std::string("X");
    result = result||character==std::string("Y");
    result = result||character==std::string("Z");
    return result;
}

bool StringHelper::IsDigit(std::string character)
{
    bool result = false;
    result = false;
    result = result||character==std::string("0");
    result = result||character==std::string("1");
    result = result||character==std::string("2");
    result = result||character==std::string("3");
    result = result||character==std::string("4");
    result = result||character==std::string("5");
    result = result||character==std::string("6");
    result = result||character==std::string("7");
    result = result||character==std::string("8");
    result = result||character==std::string("9");
    return result;
}

std::string StringHelper::ToLower(std::string input)
{
    int index = 0;
    std::string result = std::string("");
    while (index<Length(input))
    {
        std::string character = At(input,index);
        std::string lower_character = this->CharacterToLower(character);
        result = Concat(result,lower_character);
        index = index+1;
    }
    return result;
}

std::string StringHelper::CharacterToLower(std::string input)
{
    if (input==std::string("A"))
    {
        return std::string("a");
    }
    if (input==std::string("B"))
    {
        return std::string("b");
    }
    if (input==std::string("C"))
    {
        return std::string("c");
    }
    if (input==std::string("D"))
    {
        return std::string("d");
    }
    if (input==std::string("E"))
    {
        return std::string("e");
    }
    if (input==std::string("F"))
    {
        return std::string("f");
    }
    if (input==std::string("G"))
    {
        return std::string("g");
    }
    if (input==std::string("H"))
    {
        return std::string("h");
    }
    if (input==std::string("I"))
    {
        return std::string("i");
    }
    if (input==std::string("J"))
    {
        return std::string("j");
    }
    if (input==std::string("K"))
    {
        return std::string("k");
    }
    if (input==std::string("L"))
    {
        return std::string("l");
    }
    if (input==std::string("M"))
    {
        return std::string("m");
    }
    if (input==std::string("N"))
    {
        return std::string("n");
    }
    if (input==std::string("O"))
    {
        return std::string("o");
    }
    if (input==std::string("P"))
    {
        return std::string("p");
    }
    if (input==std::string("Q"))
    {
        return std::string("q");
    }
    if (input==std::string("R"))
    {
        return std::string("r");
    }
    if (input==std::string("S"))
    {
        return std::string("s");
    }
    if (input==std::string("T"))
    {
        return std::string("t");
    }
    if (input==std::string("U"))
    {
        return std::string("u");
    }
    if (input==std::string("V"))
    {
        return std::string("v");
    }
    if (input==std::string("W"))
    {
        return std::string("w");
    }
    if (input==std::string("X"))
    {
        return std::string("x");
    }
    if (input==std::string("Y"))
    {
        return std::string("y");
    }
    if (input==std::string("Z"))
    {
        return std::string("z");
    }
    return input;
}

std::string StringHelper::ToUpper(std::string input)
{
    int index = 0;
    std::string result = std::string("");
    while (index<Length(input))
    {
        std::string character = At(input,index);
        std::string upper_character = this->CharacterToUpper(character);
        result = Concat(result,upper_character);
        index = index+1;
    }
    return result;
}

std::string StringHelper::CharacterToUpper(std::string input)
{
    if (input==std::string("a"))
    {
        return std::string("A");
    }
    if (input==std::string("b"))
    {
        return std::string("B");
    }
    if (input==std::string("c"))
    {
        return std::string("C");
    }
    if (input==std::string("d"))
    {
        return std::string("D");
    }
    if (input==std::string("e"))
    {
        return std::string("E");
    }
    if (input==std::string("f"))
    {
        return std::string("F");
    }
    if (input==std::string("g"))
    {
        return std::string("G");
    }
    if (input==std::string("h"))
    {
        return std::string("H");
    }
    if (input==std::string("i"))
    {
        return std::string("I");
    }
    if (input==std::string("j"))
    {
        return std::string("J");
    }
    if (input==std::string("k"))
    {
        return std::string("K");
    }
    if (input==std::string("l"))
    {
        return std::string("L");
    }
    if (input==std::string("m"))
    {
        return std::string("M");
    }
    if (input==std::string("n"))
    {
        return std::string("N");
    }
    if (input==std::string("o"))
    {
        return std::string("O");
    }
    if (input==std::string("p"))
    {
        return std::string("P");
    }
    if (input==std::string("q"))
    {
        return std::string("Q");
    }
    if (input==std::string("r"))
    {
        return std::string("R");
    }
    if (input==std::string("s"))
    {
        return std::string("S");
    }
    if (input==std::string("t"))
    {
        return std::string("T");
    }
    if (input==std::string("u"))
    {
        return std::string("U");
    }
    if (input==std::string("v"))
    {
        return std::string("V");
    }
    if (input==std::string("w"))
    {
        return std::string("W");
    }
    if (input==std::string("x"))
    {
        return std::string("X");
    }
    if (input==std::string("y"))
    {
        return std::string("Y");
    }
    if (input==std::string("z"))
    {
        return std::string("Z");
    }
    return input;
}


};
};
};
};
};

File: S84.CTCode.Transpiler.StringHelper.ctcode.php
<?php
namespace S84\CTCode\Transpiler\StringHelper\ctcode;

include_once "S84.CTCode.System.ctcode.php";

function ClearList(array & $input): void { $input = array(); }
function Size(array $input): int { return count($input); }
function Element(array $input, int $element) { return $input[$element]; }
function Append(array & $input, mixed $element): void { $input[] = $element; }
function ClearMap(array & $input): void { reset($input); }
function SetKV(array & $input, string $key, mixed $element): void { $input[$key] = $element; }
function Keys(array $input): array { return array_keys($input); }
function HasKV(array $input, string $key): bool { return array_key_exists($key, $input); }
function GetKV(array $input, string $key) { return $input[$key]; }
function Length(string $input): int { return strlen($input); }
function At(string $input, int $index): string { return substr($input, $index, 1); }
function IntAt(string $input, int $index): int { return ord(substr($input, $index, 1)); }
function Concat(string $left, string $right): string { return $left . $right; }

class StringHelper {
    public function __construct() {
    }

    public function WriteLines(?\S84\CTCode\System\ctcode\OutputStream $destination, ?array $lines): void
    {
        $lines_index = 0;
        while ($lines_index<Size($lines))
        {
            $line = Element($lines,$lines_index);
            $destination->WriteLine($line);
            $lines_index = $lines_index+1;
        }
    }

    public function IsReserved(?string $name): ?bool
    {
        return false||$this->BeginsWith('ReservedPrefix',$name)||$this->BeginsWith('reserved_prefix_',$name)||$name=='GetType'||$name=='Object'||$name=='Return'||$name=='String'||$name=='boolean'||$name=='char'||$name=='decimal'||$name=='end'||$name=='float'||$name=='string'||false;
    }

    public function Indentation(?int $indent): ?string
    {
        $result = "";
        $result = '';
        while ($indent>0)
        {
            $indent = $indent-1;
            $result = Concat($result,'    ');
        }
        return $result;
    }

    public function BeginsWith(?string $prefix, ?string $value): ?bool
    {
        if (Length($prefix)>Length($value))
        {
            return false;
        }
        $prefix_index = 0;
        while ($prefix_index<Length($prefix))
        {
            if (At($prefix,$prefix_index)!=At($value,$prefix_index))
            {
                return false;
            }
            $prefix_index = $prefix_index+1;
        }
        return true;
    }

    public function DotToSlash(?string $input): ?string
    {
        $result = '';
        $input_index = 0;
        while ($input_index<Length($input))
        {
            $character = At($input,$input_index);
            if ($character=='.')
            {
                $result = Concat($result,'/');
            }
            else
            {
                $result = Concat($result,$character);
            }
            $input_index = $input_index+1;
        }
        return $result;
    }

    public function DotToFwdSlash(?string $input): ?string
    {
        $result = '';
        $input_index = 0;
        while ($input_index<Length($input))
        {
            $character = At($input,$input_index);
            if ($character=='.')
            {
                $result = Concat($result,'\\');
            }
            else
            {
                $result = Concat($result,$character);
            }
            $input_index = $input_index+1;
        }
        return $result;
    }

    public function StripDot(?string $input): ?string
    {
        $index = 0;
        $index = 0;
        $result = "";
        $result = '';
        while ($index<Length($input))
        {
            $character = "";
            $character = At($input,$index);
            if ($character=='.')
            {
                $result = Concat($result,'_');
            }
            else
            {
                $result = Concat($result,$character);
            }
            $index = $index+1;
        }
        return $result;
    }

    public function SnakeCaseToCamelCase(?string $snake_case): ?string
    {
        $capitalize_this_letter = true;
        $camel_case = '';
        $index = 0;
        $index = 0;
        while ($index<Length($snake_case))
        {
            $source_character = At($snake_case,$index);
            if ($source_character=='_')
            {
                $capitalize_this_letter = true;
            }
            else
            {
                if ($capitalize_this_letter==true)
                {
                    $upper_character = $this->CharacterToUpper($source_character);
                    $capitalize_this_letter = false;
                    $camel_case = Concat($camel_case,$upper_character);
                }
                else
                {
                    $capitalize_this_letter = false;
                    $camel_case = Concat($camel_case,$source_character);
                }
            }
            $index = $index+1;
        }
        return $camel_case;
    }

    public function CamelCaseToSnakeCase(?string $camel_case): ?string
    {
        $delimiter = '_';
        $snake_case = '';
        $index = 0;
        while ($index<Length($camel_case))
        {
            $source_character = At($camel_case,$index);
            $lower_character = $this->CharacterToLower($source_character);
            if ($this->IsUpper($source_character)||$this->IsDigit($source_character))
            {
                $is_first_character = Length($snake_case)==0;
                if (!$is_first_character)
                {
                    $snake_case = Concat($snake_case,$delimiter);
                }
            }
            $snake_case = Concat($snake_case,$lower_character);
            $index = $index+1;
        }
        return $snake_case;
    }

    public function IsUpper(?string $character): ?bool
    {
        $result = false;
        $result = false;
        $result = $result||$character=='A';
        $result = $result||$character=='B';
        $result = $result||$character=='C';
        $result = $result||$character=='D';
        $result = $result||$character=='E';
        $result = $result||$character=='F';
        $result = $result||$character=='G';
        $result = $result||$character=='H';
        $result = $result||$character=='I';
        $result = $result||$character=='J';
        $result = $result||$character=='K';
        $result = $result||$character=='L';
        $result = $result||$character=='M';
        $result = $result||$character=='N';
        $result = $result||$character=='O';
        $result = $result||$character=='P';
        $result = $result||$character=='Q';
        $result = $result||$character=='R';
        $result = $result||$character=='S';
        $result = $result||$character=='T';
        $result = $result||$character=='U';
        $result = $result||$character=='V';
        $result = $result||$character=='W';
        $result = $result||$character=='X';
        $result = $result||$character=='Y';
        $result = $result||$character=='Z';
        return $result;
    }

    public function IsDigit(?string $character): ?bool
    {
        $result = false;
        $result = false;
        $result = $result||$character=='0';
        $result = $result||$character=='1';
        $result = $result||$character=='2';
        $result = $result||$character=='3';
        $result = $result||$character=='4';
        $result = $result||$character=='5';
        $result = $result||$character=='6';
        $result = $result||$character=='7';
        $result = $result||$character=='8';
        $result = $result||$character=='9';
        return $result;
    }

    public function ToLower(?string $input): ?string
    {
        $index = 0;
        $result = '';
        while ($index<Length($input))
        {
            $character = At($input,$index);
            $lower_character = $this->CharacterToLower($character);
            $result = Concat($result,$lower_character);
            $index = $index+1;
        }
        return $result;
    }

    public function CharacterToLower(?string $input): ?string
    {
        if ($input=='A')
        {
            return 'a';
        }
        if ($input=='B')
        {
            return 'b';
        }
        if ($input=='C')
        {
            return 'c';
        }
        if ($input=='D')
        {
            return 'd';
        }
        if ($input=='E')
        {
            return 'e';
        }
        if ($input=='F')
        {
            return 'f';
        }
        if ($input=='G')
        {
            return 'g';
        }
        if ($input=='H')
        {
            return 'h';
        }
        if ($input=='I')
        {
            return 'i';
        }
        if ($input=='J')
        {
            return 'j';
        }
        if ($input=='K')
        {
            return 'k';
        }
        if ($input=='L')
        {
            return 'l';
        }
        if ($input=='M')
        {
            return 'm';
        }
        if ($input=='N')
        {
            return 'n';
        }
        if ($input=='O')
        {
            return 'o';
        }
        if ($input=='P')
        {
            return 'p';
        }
        if ($input=='Q')
        {
            return 'q';
        }
        if ($input=='R')
        {
            return 'r';
        }
        if ($input=='S')
        {
            return 's';
        }
        if ($input=='T')
        {
            return 't';
        }
        if ($input=='U')
        {
            return 'u';
        }
        if ($input=='V')
        {
            return 'v';
        }
        if ($input=='W')
        {
            return 'w';
        }
        if ($input=='X')
        {
            return 'x';
        }
        if ($input=='Y')
        {
            return 'y';
        }
        if ($input=='Z')
        {
            return 'z';
        }
        return $input;
    }

    public function ToUpper(?string $input): ?string
    {
        $index = 0;
        $result = '';
        while ($index<Length($input))
        {
            $character = At($input,$index);
            $upper_character = $this->CharacterToUpper($character);
            $result = Concat($result,$upper_character);
            $index = $index+1;
        }
        return $result;
    }

    public function CharacterToUpper(?string $input): ?string
    {
        if ($input=='a')
        {
            return 'A';
        }
        if ($input=='b')
        {
            return 'B';
        }
        if ($input=='c')
        {
            return 'C';
        }
        if ($input=='d')
        {
            return 'D';
        }
        if ($input=='e')
        {
            return 'E';
        }
        if ($input=='f')
        {
            return 'F';
        }
        if ($input=='g')
        {
            return 'G';
        }
        if ($input=='h')
        {
            return 'H';
        }
        if ($input=='i')
        {
            return 'I';
        }
        if ($input=='j')
        {
            return 'J';
        }
        if ($input=='k')
        {
            return 'K';
        }
        if ($input=='l')
        {
            return 'L';
        }
        if ($input=='m')
        {
            return 'M';
        }
        if ($input=='n')
        {
            return 'N';
        }
        if ($input=='o')
        {
            return 'O';
        }
        if ($input=='p')
        {
            return 'P';
        }
        if ($input=='q')
        {
            return 'Q';
        }
        if ($input=='r')
        {
            return 'R';
        }
        if ($input=='s')
        {
            return 'S';
        }
        if ($input=='t')
        {
            return 'T';
        }
        if ($input=='u')
        {
            return 'U';
        }
        if ($input=='v')
        {
            return 'V';
        }
        if ($input=='w')
        {
            return 'W';
        }
        if ($input=='x')
        {
            return 'X';
        }
        if ($input=='y')
        {
            return 'Y';
        }
        if ($input=='z')
        {
            return 'Z';
        }
        return $input;
    }

}

?>
